### ISSUE TITLE
Decode performance optimizations for MLX-LM inference

### ISSUE THREAD

**Issue by dahai80** [ml-explore/mlx-lm#1450]
## Decode Performance Optimizations for MLX-LM Inference

We have implemented and benchmarked a set of decode-time optimizations for MLX-LM on Apple Silicon. On an M2 Ultra 137GB, these combine to deliver up to **162% faster decode throughput** over the baseline mxfp8 quantization. We are proposing these for upstream integration into mlx-lm.

---

### 1. Mixed-bit Quantization Recipes

**Problem:** Uniform quantization wastes bits on layers that tolerate aggressive quantization and over-quantizes sensitive layers.

**Proposal:** Pre-tuned per-layer bit allocation plans (`mixed_3_4`, `mixed_2_6`, `quant2_all`, etc.) that apply different bit widths to different layer groups based on their sensitivity. Exposed via a `--quant-recipe` flag in `mlx_lm.convert`.

Example usage:
```bash
python -m mlx_lm.convert --model Qwen/Qwen2.5-27B --quant-recipe mixed_3_4
```

Recipes assign bit widths to layer groups (embeddings, attention projections, MLP projections, output head) based on empirically measured sensitivity. This avoids the need for per-model calibration while still capturing most of the mixed-precision benefit.

---

### 2. Skip-logsumexp in Decode

**Problem:** During autoregressive decode, only one new token position is computed per step. The full logsumexp normalization over the vocabulary is unnecessary for correctness when sampling from a single-token logit vector.

**Proposal:** Skip the logsumexp normalization in the decode path. Since we only sample one token, the relative ordering of logits is preserved without normalization, and the sampling step (argmax or random) produces identical results.

---

### 3. Fused Sampler

**Problem:** Sampling (argmax for greedy, or random sampling with temperature/top-p/top-k) is done in Python with multiple dispatches to Metal — one to apply temperature, one for top-p masking, one for the random draw, etc.

**Proposal:** Combine the entire sampling operation into a single Metal kernel. This eliminates multiple Python-to-Metal round trips per token and reduces kernel launch overhead.

---

### 4. FP8 KV Cache

**Problem:** The KV cache is stored in float16, consuming 2 bytes per element. For long sequences and large models, this dominates memory bandwidth during attention.

**Proposal:** Store the KV cache in E4M3 FP8 format (1 byte per element), halving the memory bandwidth for attention. The FP8 values are upcast to float16 on load inside the attention kernel. Quality impact is minimal because KV cache values have a narrow dynamic range that FP8 E4M3 represents well.

---

### 5. mx.compile for Decode

**Problem:** The single-token decode step involves many small Metal operations that are re-dispatched every step, even though the graph structure is identical across steps.

**Proposal:** Apply `mx.compile` to the decode step. Since the decode graph is fixed (same ops, same shapes every step), compilation eliminates Python dispatch overhead and allows Metal to optimize the kernel sequence.

---

### 6. Python Dispatch Overhead Reduction

**Problem:** The per-token Python overhead in the decode loop (function calls, dict lookups, conditionals) adds up and becomes a significant fraction of total step time at high throughput.

**Proposal:** Streamline the decode loop — hoist invariants out of the loop, flatten dispatch paths, and reduce per-step Python work to the minimum needed for correctness.

---

### Benchmark Results

All measurements on **Qwen3.6-27B on M2 Ultra 137GB**, decode tokens/sec:

| Recipe | BPW | Speed (tok/s) | vs mxfp8 |
|--------|-----|---------------|----------|
| mxfp8 | 8.0 | 18.5 | baseline |
| mxfp4 | 4.0 | 32.3 | +75% |
| mixed_3_4 | 3.68 | 36.2 | +96% |
| mixed_2_6 | 3.25 | 39.3 | +112% |
| quant2_all | 2.37 | 48.5 | +162% |

The speedup comes from two multiplicative factors: reduced memory bandwidth (lower BPW means less data to load per layer) and the decode-specific optimizations (skip-logsumexp, fused sampler, mx.compile, dispatch reduction) which improve throughput at any quantization level.

---

### Proposed Integration Path

1. **Mixed-bit recipes** — Add `--quant-recipe` to `mlx_lm.convert` with the built-in recipe definitions. Low risk, purely a conversion-time change.
2. **FP8 KV cache** — Add as an opt-in flag (e.g. `--fp8-kv-cache`). Requires a small change to the attention implementation.
3. **mx.compile for decode** — Wrap the decode step in `mx.compile`. Should be a one-line change in the generate loop.
4. **Skip-logsumexp** — Conditional in the sampler: skip normalization when in decode mode.
5. **Fused sampler** — New Metal kernel. Larger change, but self-contained.
6. **Dispatch reduction** — Refactor the decode loop. Incremental, can be done in multiple PRs.

Happy to open individual PRs for each optimization if there is interest.
